### ISSUE TITLE
The model size and the amount of parameters decreased after pruning, but the inference time did not change significantly  

### ISSUE THREAD

**Issue by yebuxiu123** [ultralytics/ultralytics#5238]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I'm working on the inference speed test related to the yolov8s-seg model, and I've performed the inference using the original model (size 23.8MB), the improved model with module replacement (size 18.1MB), and the pruned model based on the improved model (size 10.2MB), using the command 

yolo task=segment mode=val model=runs/segment/me/train14_s_e200_c60_yaml/weights/best.pt data=./coco128-seg.yaml  batch=1 workers=0

But the test speed does not meet my expectations, I replace the lightweight module to improve the model for 18.1MB, compared to the original model volume reduced by nearly 20%, but the test inference time is the same as 9.2ms, after the pruning of the model size of 10.2MB, the inference time for the 8.4ms，the following is the result of the reasoning, from top to bottom this is the original model, the improved model, the pruned model. It can be seen that the number of model parameters has been reduced accordingly, and the number of parameters of the pruned model is about half of the original model, but why the inference time is not reduced accordingly with the reduction of the model volume and the number of parameters? Or is there something wrong with the way I measure the model fps?

Ultralytics YOLOv8.0.139 🚀 Python-3.8.13 torch-1.10.1 CUDA:0 (NVIDIA GeForce RTX 2060, 5926MiB)
YOLOv8s-seg summary (fused): 195 layers, 11779987 parameters, 0 gradients
val: Scanning /media/gc/1T/ultralytics-main/1/data/labels/val.cache... 373 images, 0 backgrounds, 0 corr
  Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P     
    all        373        463      0.883       0.77      0.862      0.657      0.832      0.717      0.767      0.305
Speed: 0.3ms preprocess, 9.2ms inference, 0.0ms loss, 0.7ms postprocess per image
Results saved to runs/segment/val19

Ultralytics YOLOv8.0.139 🚀 Python-3.8.13 torch-1.10.1 CUDA:0 (NVIDIA GeForce RTX 2060, 5926MiB)
YOLOv8s_Ghost_Gsconv_LSKblock1-seg summary: 291 layers, 8891977 parameters, 0 gradients
val: Scanning /media/gc/1T/ultralytics-main/1/data/labels/val.cache... 373 images, 0 backgrounds, 0 corr
  Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P     
    all        373        463      0.928      0.793      0.882      0.678      0.862       0.73      0.768      0.297
Speed: 0.3ms preprocess, 9.2ms inference, 0.0ms loss, 0.7ms postprocess per image
Results saved to runs/segment/val20

Ultralytics YOLOv8.0.139 🚀 Python-3.8.13 torch-1.10.1 CUDA:0 (NVIDIA GeForce RTX 2060, 5926MiB)
YOLOv8s_Ghost_Gsconv_LSKblock1-seg summary: 291 layers, 5456107 parameters, 0 gradients
val: Scanning /media/gc/1T/ultralytics-main/1/data/labels/val.cache... 373 images, 0 backgrounds, 0 corr
  Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P     
    all        373        463      0.918      0.801      0.866      0.671      0.842      0.747      0.753      0.302
Speed: 0.4ms preprocess, 8.4ms inference, 0.0ms loss, 0.8ms postprocess per image
Results saved to runs/segment/val18

**Comment by RizwanMunawar** [comment]
Hello @yebuxiu123, Thanks for the detailed Information about the issue you are facing. It's important to note that there are two primary types of pruning in neural networks: neuron pruning and weight pruning.

- Neuron pruning removes entire neurons, while weight pruning eliminates specific weight parameters in a neural network, both aiming to reduce model size and complexity.

In your scenario, it appears that you're employing neuron pruning, which should indeed reduce inference time. However, it's important to conduct actual testing to assess the impact accurately. Several factors can influence the results. It would be greatly appreciated if you could provide both the pruned model file and the original model file so that we can perform internal testing and provide you with detailed feedback.

Thanks

**Comment by yebuxiu123** [comment]
Very glad to see your reply, I actually perform trimming by removing unimportant channels through the gamma coefficient of the model's BN layer, and I'm not quite sure which kind of trimming that counts as. Also, I'd be happy to provide my training weights, how do I provide it? Sent to your e-mail?

**Comment by glenn-jocher** [comment]
@yebuxiu123 hello, 

Thank you for further explaining your pruning process. By using the auxiliary variable gamma coefficient in the batch normalization layer to guide the pruning process, effectively you're performing a form of channel pruning. Channel pruning aims to prune unimportant channels in the convolutional layer, which is beneficial in both reducing model complexity and improving inference speed.

About sharing the weights, you can upload the files to a cloud storage of your preference such as Google Drive or Dropbox. After uploading, you can share the shareable link via GitHub issue or direct message.

As for the model inference speed (fps), the value you mentioned is indeed a good way to gauge it. The inference speed is typically measured by the time taken for the model to produce predictions. Your command seems to be calculating time on inference, and therefore, outputting a reasonable estimate of speed.

However, bear in mind there can be minor variations based on factors such as the size of the batch being processed, hardware specifications, software optimizations, among others, so it's advisable to conduct multiple tests and average the results for a more accurate measure.

Hope this helps, let me know if you have any further questions.

**Comment by yebuxiu123** [comment]
I uploaded the three model weights mentioned earlier to a Google cloud drive at the link 

https://drive.google.com/file/d/1VWSVM7vFwhRxolgG391xNdtOUFZ_l5sO/view?usp=drive_link

This includes the weights trained using the original yolov8s, the weights from the improved lightweighting, and the weights from pruning the improved model, in addition to containing my ultralytics folder, which contains some of the code for modules I've added that may need to be loaded when testing my model weights. I hope you get the desired model inference speed test results

**Comment by glenn-jocher** [comment]
Hello @yebuxiu123,

Thanks for providing the model weights, it'll be a great help in investigating the situation further. From what you've described, it sounds like you've done a thorough job in preparing and pruning your models.

When we test these models, we'll consider various factors that could affect the inference speed, such as hardware specifications, software optimizations, and batch processing size. This will help us to assess whether the pruning has led to the expected improvements in speed.

Batch size is an important factor, as smaller batches can take longer to process per item, but would use less overall memory, and vice-versa for larger batches. We'll take into account factors such as this during our testing.

Moreover, our testing will be around how long the trained model takes to make predictions on unseen data, which is an accurate measure of inference speed.

Your pruned model is expected to have a faster inference speed since it has fewer parameters compared to the original model. However, the benefits from model pruning can sometimes be offset by sub-optimal hardware usage or certain software inefficiencies. We will look into this to ensure the test results are both optimal and accurate.

Thank you for your patience and collaboration. Please note that the process might take some time, but be assured that your request is being attended to. We'll update you with our findings as soon as we complete the tests.

**Comment by yebuxiu123** [comment]
Thank you very much for your prompt reply, it occurs to me that my model was trained on some images of road cracks in visible vs. infrared, do you need me to provide a portion of the test images if you are doing test work on my trained model model?

**Comment by glenn-jocher** [comment]
Hello @yebuxiu123, 

Thanks for your message and your offer. Yes, providing a few of your test images that you have used while training your models would be very helpful. By using the same test images, we can ensure that the test environment matches your training environment as closely as possible, which could provide us with more accurate and relevant results.

Thanks again for your cooperation and patience.

**Comment by yebuxiu123** [comment]
I uploaded the validation data I used to Google Cloud Drive, which contains images and txt labeled data, please feel free to ask if there is anything else you would like me to provide as you test your model!
https://drive.google.com/file/d/1qFr7nS-KXT4T68Vj5LGwxOTcSli6e7M1/view?usp=drive_link

**Comment by glenn-jocher** [comment]
Hello @yebuxiu123,

Thank you for uploading your validation dataset. It will be immensely helpful in providing us with more accurate and representative test results. As per your shared models and this validation dataset, we will set up a testing scenario that closely mirrors your original setup. Rest assured, if we need any more information or data from your end, we will let you know promptly. 

We appreciate your cooperation. This detailed data will allow us to perform a comprehensive evaluation. Once we have the results, we will promptly share our findings with you. 

Thanks again for your patience and support in this process.

**Comment by sabillahm** [comment]
Hi. I have the same problem. The model size and the amount of parameters decreased after pruning, but the inference time did not change significantly. Do you have references such as scientific articles that said "various factors that could affect the inference speed, such as hardware specifications, software optimizations, and batch processing size. model pruning can sometimes be offset by sub-optimal hardware usage or certain software inefficiencies"

I need your help

**Comment by glenn-jocher** [comment]
Hello, thank you for reaching out. While pruning often reduces model size and parameters, inference time may not always decrease significantly due to factors like hardware utilization, software optimizations, or specific pruning techniques. For a deeper understanding, you can explore [Ultralytics' Model Pruning documentation](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity/) or refer to research on pruning techniques and their impact on efficiency. Let us know if you have further questions!
